Optimize common newline regex

nipunshukla21 opened the pull request:
#### cbb2ba2c0577f5eca058938242c4ab363130c56d
<pre>
Optimize common newline regex
<a href="https://bugs.webkit.org/show_bug.cgi?id=302014">https://bugs.webkit.org/show_bug.cgi?id=302014</a>
<a href="https://rdar.apple.com/163958152">rdar://163958152</a>

Reviewed by Yusuke Suzuki.

The regular expression \r\n?|\n is a very common syntax for newline splitting.
This patch adds an entry for this to Yarr::SpecificPattern and uses SIMD to further optimize
this case.

* LayoutTests/fast/regex/script-tests/string-split-newline.js: Added.
(newlinePatterns.forEach):
* LayoutTests/fast/regex/string-split-newline-expected.txt: Added.
* LayoutTests/fast/regex/string-split-newline.html: Added.
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::tryTrimSpaces):
(JSC::replaceUsingRegExpSearch):
* Source/JavaScriptCore/yarr/Yarr.h:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::extractSpecificPattern):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::findNextNewline):

Canonical link: <a href="https://commits.webkit.org/302933@main">https://commits.webkit.org/302933@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d64eef496a02fb27d3bc08026873bf2bb393c2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130553 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41508 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137971 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82166 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2716 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99464 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2064 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116907 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80169 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1985 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35039 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81230 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122556 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110553 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140450 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129006 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2382 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107975 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2658 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107895 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27477 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2024 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31685 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55592 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2684 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66073 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162021 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2503 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40399 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2705 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2610 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -623,6 +624,73 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
+                if (resultLength >= limit)
+                    break;
+
+                position = newlinePos.position + newlinePos.length;
```
should this regex return empty strings in the case that there are several \n chars consecutively?

nipunshukla21 replied:
Yes, this is the output for both release Safari and a locally built copy with my changes:
```
> var newlinePattern = /\r\n?|\n/;
< undefined
> "\n\n\n\n\n\n\n".split(newlinePattern)
< ["", "", "", "", "", "", "", ""] (8)
> "\r\n\r\n\r\n\r\n\r\n\r\n\r\n".split(newlinePattern)
< ["", "", "", "", "", "", "", ""] (8)
> "\r\r\r\r\r\r\r".split(newlinePattern)
< ["", "", "", "", "", "", "", ""] (8)
```
That's not to say there aren't other edge cases missed here though!

Andrew-Fryer commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -1202,6 +1202,55 @@ inline size_t countMatchedCharacters(std::span<const CharacterType> span, Charac
+struct NewlinePosition {
+    size_t position { notFound };
+    size_t length { 0 };
+};
```
IMHO this is good for readability, but I think this isn't really idiomatic WTF, so we should definitely get an r+ from someone who has opinions here such as Chris Dumez.

Andrew-Fryer commented on LayoutTests/fast/regex/script-tests/string-split-newline.js:
```diff
@@ -0,0 +1,27 @@
+shouldBe('"abc".split(newlinePattern)', '["abc"]');
+shouldBe('"a\\nb\\nc".split(newlinePattern)', '["a", "b", "c"]');
+shouldBe('"a\\nb\\n".split(newlinePattern, 2)', '["a", "b"]');
+shouldBe('"a\\nb\\nc\\n".split(newlinePattern, 3)', '["a", "b", "c"]');
```
There's a nice testing idiom that might work well here. You just have some examples and then the test runs them through your code with the optimization on and off and then diffs the results.
That said, I think it's fine to test the way this is already written.

danlliu replied:
For completeness, can we also add a test which splits a string longer than 32 characters? This ensures that `vectorMatch` is also tested for correctness.

Andrew-Fryer commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -2346,11 +2346,70 @@ class YarrPatternConstructor {
             return;
+
+        if (tryExtractNewlines())
+            return;
```
I think it might make sense to either make this check more generic (so it passes `\n|\r\n?` for example) or do the check directly on the string passed into the regex function so that the match is simpler and happens earlier.

Constellation commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -1202,6 +1202,55 @@ inline size_t countMatchedCharacters(std::span<const CharacterType> span, Charac
+        return { };
+
+    size_t pos = ptr - searchSpan.data();
+    CharacterType ch = searchSpan[pos];
```
Let's just have `CharacterType ch = *ptr;`

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -623,6 +624,73 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
+        if (!limit)
+            RELEASE_AND_RETURN(scope, JSValue::encode(constructEmptyArray(globalObject, nullptr)));
+
+        unsigned inputSize = input->length();
```
This part is the same for both non specific pattern and others. Can we put it outside of if statement?

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -623,6 +624,73 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
+        if (!inputSize) {
+            JSArray* result = constructEmptyArray(globalObject, nullptr);
+            RETURN_IF_EXCEPTION(scope, { });
+            result->putDirectIndex(globalObject, 0, inputString);
```
RegExp matching is also changing RegExpCache state. So you need to make sure the cache state (`RegExp.$1` etc.) is the same to when we are doing matching.

Constellation replied:
And can you add a test for that?

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -623,6 +624,73 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
+                auto newlinePos = WTF::findNextNewline(span, position);
+
+                if (newlinePos.position == WTF::notFound) {
+                    if (position < inputSize) {
+                        result->putDirectIndex(globalObject, resultLength++, jsSubstringOfResolved(vm, inputString, position, inputSize - position));
+                        RETURN_IF_EXCEPTION(scope, AbortSplit);
+                    }
+                    break;
+                }
+
+                result->putDirectIndex(globalObject, resultLength++, jsSubstringOfResolved(vm, inputString, position, newlinePos.position - position));
+                RETURN_IF_EXCEPTION(scope, AbortSplit);
+
+                if (resultLength >= limit)
+                    break;
+
+                position = newlinePos.position + newlinePos.length;
+            }
+            return ContinueSplit;
+        };
```
Ditto about `RegExp.$1` etc. state (RegExpCache state).

Constellation replied:
Can you add a test for this too?

Ahmad-S792 commented on LayoutTests/fast/regex/string-split-newline.html:
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
+<html>
+<head>
+<script src="../../resources/js-test-pre.js"></script>
```
You can just use `js-test.js` rather than `pre` and `post`.

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -667,6 +668,55 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
+            processSplit(input->span8());
+        else
+            processSplit(input->span16());
+
```
Remove this empty line to make processSplit and `RETURN_IF_EXCEPTION` tied.

Constellation commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -667,6 +668,55 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
+        auto processSplit = [&](auto span) {
+            while (position < inputSize && resultLength < limit) {
+                auto newlinePos = WTF::findNextNewline(span, position);
+
```
Remove this empty line to make `newlinePos.position == WTF::notFound` check tied to the above visually.

nipunshukla21 committed:
Optimize common newline regex
https://bugs.webkit.org/show_bug.cgi?id=302014
rdar://163958152

Reviewed by Yusuke Suzuki.

The regular expression \r\n?|\n is a very common syntax for newline splitting.
This patch adds an entry for this to Yarr::SpecificPattern and uses SIMD to further optimize
this case.

* LayoutTests/fast/regex/script-tests/string-split-newline.js: Added.
(newlinePatterns.forEach):
* LayoutTests/fast/regex/string-split-newline-expected.txt: Added.
* LayoutTests/fast/regex/string-split-newline.html: Added.
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::tryTrimSpaces):
(JSC::replaceUsingRegExpSearch):
* Source/JavaScriptCore/yarr/Yarr.h:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::extractSpecificPattern):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::findNextNewline):

Canonical link: https://commits.webkit.org/302933@main
